File: apps/android/app/src/main/java/com/kungsbackacarcommunity/app/chatchannels/CommunityChatRepository.kt
package com.kungsbackacarcommunity.app.chatchannels

import com.kungsbackacarcommunity.app.chat.ChatReportReason
import kotlinx.coroutines.flow.Flow

/**
 * Community (app-wide) chat access. Reads are a live Firestore listener on
 * `communityChat/global/messages` (rules grant any active member the read);
 * sending, older-page pagination, and mark-read go through the member-gated
 * `communityChat-*` callables (europe-west1). Firebase-free interface so the
 * coordinator/screens are unit- and UI-testable with fakes (mirrors DmRepository).
 *
 * Unread is a lightweight per-user LAST-READ marker
 * (`userPrivate/{uid}.communityChatLastReadAt`, owner-only readable) rather than a
 * fan-out counter: [observeUnread] combines the newest-message listener with the
 * marker so the caller sees a dot when the newest message post-dates their marker.
 */
interface CommunityChatRepository {
    /** Live newest-window of the community channel, chronological. */
    fun observeMessages(): Flow<ChannelMessagesState>

    /**
     * Live "has unread" flag for [uid]: true while the newest message is from
     * someone else and newer than the caller's last-read marker. Emits false
     * once [markRead] runs (or while the channel is empty).
     */
    fun observeUnread(uid: String): Flow<Boolean>

    /**
     * `communityChat-post` — posts [text] to the global channel, @mentioning
     * [mentionedUids] (at most [MAX_MESSAGE_MENTIONS] — the server REJECTS more
     * with `invalid-argument`). The uids come from the composer's @-picker, never
     * from parsing names out of [text]: display names are not unique, so a parsed
     * mention would sooner or later notify the wrong member. The returned
     * [ChannelSendResult.Sent] echoes the ACCEPTED subset.
     *
     * [clientId] is the optimistic idempotency key: the backend uses it as the
     * message doc id, so a retry is exactly-once and the client reconciles its
     * bubble by matching it. Null falls back to a server auto-id (legacy) doc.
     */
    suspend fun post(
        text: String,
        mentionedUids: List<String> = emptyList(),
        clientId: String? = null,
    ): ChannelSendResult

    /**
     * `communityChat-post` with an inline reply target. [replyToMessageId] is the
     * id of the message being replied to (WhatsApp-style quote, not a thread): the
     * client sends only the id and the server snapshots the parent from THIS
     * channel itself (ignored while the `chatReplies` flag is off). Null is an
     * ordinary, non-reply message.
     *
     * Defaults to delegating to the 3-arg overload, so a fake predating replies
     * keeps working (it drops the reply target); the Firebase repository overrides
     * this to forward the id.
     */
    suspend fun post(
        text: String,
        mentionedUids: List<String>,
        clientId: String?,
        replyToMessageId: String?,
    ): ChannelSendResult = post(text, mentionedUids, clientId)

    /** `communityChat-list` — an older page before the [before] ISO cursor. */
    suspend fun loadOlder(before: String): ChannelOlderResult

    /** `communityChat-markRead` — stamps the caller's last-read marker. Idempotent. */
    suspend fun markRead()

    /**
     * `chatchannels-reportMessage` (channel: 'community') — files a moderation
     * report against the message [messageId] with [reason]. Eligibility mirrors the
     * channel read rule (any active member), reporting your OWN message is rejected
     * server-side (the client also gates this on [MessageModeration.canActOn]), and
     * the message is snapshotted into the report so a later TTL-delete can't blank
     * it. The reported message id is a community-channel doc id; the channel scope
     * ('global') is fixed server-side, so the client sends only the reason + id.
     *
     * Binary [ChannelReportResult] by design: the queue is fire-and-forget and the
     * backend never reveals whether a prior report existed, so the reporter only
     * learns "reached the backend" vs "didn't".
     *
     * Defaults to [ChannelReportResult.Failed] so a fake predating reporting fails
     * closed rather than falsely reporting success; the Firebase repository overrides it.
     */
    suspend fun report(messageId: String, reason: ChatReportReason): ChannelReportResult =
        ChannelReportResult.Failed
}

File: apps/ios/KCC/Chat/CommunityChatRepository.swift
import Foundation

/// The single APP-WIDE community channel's read/write access — the iOS port of
/// Android's `CommunityChatRepository`. Firebase-free so the coordinator and
/// screens are unit-testable with fakes (mirrors ``EventsRepository``).
///
/// Reads are a direct live Firestore snapshot listener on
/// `communityChat/global/messages` (member-readable per
/// firebase/firestore.rules), bounded to ``channelMessagesPageSize`` newest-first
/// and filtered CLIENT-side against the caller's mutual-hidden set
/// (``ChatBlockVisibility``) — a Firestore rule cannot filter a list query
/// per-document. Sending, older-page pagination, mark-read, and reporting go
/// through the member-gated `communityChat-*` / `chatchannels-*` callables
/// (europe-west1); the client never writes the message tree.
protocol CommunityChatRepository: AnyObject, Sendable {
    /// Live newest-window of the community channel, already block-filtered and
    /// chronological (oldest-first). Each call returns a fresh stream backed by
    /// its own listeners; terminating the stream detaches them. Emits SETTLED
    /// states only — the coordinator supplies ``ChannelMessagesState/loading``
    /// before the first emission. Android: `observeMessages`.
    func observeMessages() -> AsyncStream<ChannelMessagesState>

    /// An older page before the `before` ISO-8601 cursor (`communityChat-list`,
    /// page size 30). Android: `loadOlder`.
    func loadOlder(before: String) async -> ChannelOlderResult

    /// `communityChat-post` — posts `text` with the accepted @mention set.
    /// `clientId` is the optimistic idempotency key (used verbatim as the doc
    /// id). `replyToMessageId` is sent only when the `chatReplies` flag is on
    /// (the backend ignores it otherwise). Android: `post`.
    func post(
        text: String,
        mentionedUids: [String],
        clientId: String?,
        replyToMessageId: String?
    ) async -> ChannelSendResult

    /// `communityChat-markRead` — best-effort; a transient failure is swallowed.
    /// Android: `markRead`.
    func markRead() async

    /// `chatchannels-reportMessage` with `channel: "community"`. Android:
    /// `report`.
    func report(messageId: String, reason: ChatReportReason) async -> ChannelReportResult

    /// The signed-in user's uid, or nil with no session — the Firebase seam owns
    /// identity so coordinators stay Firebase-free (as in ``EventsRepository``).
    func currentUserId() -> String?
}
